Create GeoSeries and GeoDataFrame subclasses from polars in Python

kylebarron opened the pull request:
Closes #33.

The goal of this is to create a similar API to geopandas, where it subclasses `Series` and `DataFrame`, assigning methods directly onto those classes.

Questions:

- What's the best way to return an instance of `GeoDataFrame` and `GeoSeries`? Ideally, the actual data representation contained in each of these won't be different; it'll just be the python object that additionally has more methods. The pure functions in the python-rust bindings currently return a bare polars Series, so we could:
    - Just wrap the return object in a call to `GeoSeries(...)` in Python, i.e.

        ```py
        def centroid(self) -> GeoSeries:
            return GeoSeries(geopolars.centroid(self))
        ```

    - Try to make this conversion automatically from the rust side? The current rust bindings call `polars.from_arrow` to reconstruct a polars series: 
https://github.com/kylebarron/geopolars/blob/71cd2027617324a15c609450ec90fafe3b90213c/py-geopolars/src/ffi.rs#L92 Could we just create a `geopolars.from_arrow` function that returns a `GeoSeries`/`GeoDataFrame`? I think this might actually be easier, as long as we don't run into any weird circular dependency issues.

kylebarron commented on py-geopolars/src/ffi.rs:
```diff
@@ -1,5 +1,5 @@
 use arrow::{array::ArrayRef, ffi};
-use polars::prelude::*;
+use polars::prelude::{ArrowField, Series};
```
This is just a nit that I don't love imports into the global namespace. Especially for things like `ArrowField` that could have come from `arrow2`

kylebarron replied:
oh no you didn't add that! I was just explaining the diff 🙂

kylebarron commented:
I don't know why there are obscure [`cargo metadata` errors](https://github.com/kylebarron/geopolars/runs/6874508314?check_suite_focus=true#step:5:11) (only on CI) when adding a python package to the top-level workspace. (note that the error is about `geopolars/geopolars/py-geopolars`, when that folder doesn't exist)

But the python libs also aren't in the polars workspace: https://github.com/pola-rs/polars/blob/4154c6af37cc366868289059f70d5146d72d5848/Cargo.toml#L2-L13

stuartlynn commented:
> Closes #33.
> 
> The goal of this is to create a similar API to geopandas, where it subclasses `Series` and `DataFrame`, assigning methods directly onto those classes.
> 
> Questions:
> 
> * What's the best way to return an instance of `GeoDataFrame` and `GeoSeries`? Ideally, the actual data representation contained in each of these won't be different; it'll just be the python object that additionally has more methods. The pure functions in the python-rust bindings currently return a bare polars Series, so we could:
>   
>   * Just wrap the return object in a call to `GeoSeries(...)` in Python, i.e.
>     ```python
>     def centroid(self) -> GeoSeries:
>         return GeoSeries(geopolars.centroid(self))
>     ```
>   * Try to make this conversion automatically from the rust side? The current rust bindings call `polars.from_arrow` to reconstruct a polars series:
>     https://github.com/kylebarron/geopolars/blob/71cd2027617324a15c609450ec90fafe3b90213c/py-geopolars/src/ffi.rs#L92
>      Could we just create a `geopolars.from_arrow` function that returns a `GeoSeries`/`GeoDataFrame`? I think this might actually be easier, as long as we don't run into any weird circular dependency issues.

I think my preference here would be to make this as automatic as we can, esp for future contributors who might not catch the need to return the series wrapped as a GeoSeries. 

One small concern though is that we will have times when the API is actually returning a Series, rather than a GeoSeries. For example the method that calculates the area. Should we have two functions then? One for rust_series_to_py_series and one rust_series_to_py_geoseries ? We presumably wouldn't want the python series to have the Geo functions if it's a regular series? 

Perhaps instead of having two functions we can make the rust_series_to_py_series inspect the series it is passed to see if it's a geoseries and conditionally wrap the series in either a vanilla Series or a GeoSeries

kylebarron commented:
I should have some time tomorrow afternoon to finish up this PR

kylebarron commented:
I added a `from_arrow` function in `convert.py` (mirroring the layout of `py-polars`) and added  `rust_series_to_py_geoseries` in addition to `rust_series_to_py_series` in ffi.rs.

kylebarron commented:
I added a `from_arrow` function in `convert.py` (mirroring the layout of `py-polars`) and added  `rust_series_to_py_geoseries` in addition to `rust_series_to_py_series` in ffi.rs.
